src/librustc/traits/error_reporting.rs

```diff
@@ -523,6 +524,11 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     {
         let span = obligation.cause.span;
 
+        if !self.reported_selection_errors.borrow_mut().insert((span, error.clone())) {
```

Comment:
This sounds overly excessive. Outside of closures all trait errors are `Unimplemented`, so you are only showing 1 error per span.